File: QytetetJava/src/modeloqytetet/Calle.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author fran
 */
public class Calle extends Casilla {

    private TituloPropiedad titulo;
    private int numHoteles;
    private int numCasas;

    public Calle(int numeroCasilla, int coste, TipoCasilla tipo, TituloPropiedad titulo) {
        super(numeroCasilla, coste, tipo);
        this.numHoteles = 0;
        this.numCasas = 0;
        setTitulo(titulo);

    }

    public TituloPropiedad getTitulo() {
        return titulo;
    }

    public int getNumHoteles() {
        return numHoteles;
    }

    public int getNumCasas() {
        return numCasas;
    }

    public int getCosteHipoteca() {
        throw new UnsupportedOperationException("Sin implementar");
    }

    public int getPrecioEdificar() {
        return titulo.getPrecioEdificar();
    }

    private void setTitulo(TituloPropiedad titulo) {
        this.titulo = titulo;
        titulo.setCalle(this);
    }

    void setNumHoteles(int numHoteles) {
        this.numHoteles = numHoteles;
    }

    void setNumCasas(int numCasas) {
        this.numCasas = numCasas;
    }

    TituloPropiedad asignarPropietario(Jugador jugador) {
        titulo.setPropietario(jugador);
        return titulo;
    }

    int calcularValorHipoteca() {
        int hipotecaBase = titulo.getHipotecaBase();
        return hipotecaBase + (int) (numCasas * 0.5 * hipotecaBase + numHoteles * hipotecaBase);
    }

    int cancelarHipoteca() {
        titulo.setHipotecada(false);
        int precio = (int) (1.1 * calcularValorHipoteca());
        return precio;
    }

    int cobrarAlquiler() {
        int costeAlquilerBase = titulo.getAlquilerBase();
        int costeAlquiler = costeAlquilerBase + (int) (numCasas * 0.5 + numHoteles * 2);
        titulo.cobrarAlquiler(costeAlquiler);
        return costeAlquiler;
    }

    int edificarCasa() {
        setNumCasas(numCasas + 1);
        return titulo.getPrecioEdificar();
    }

    int edificarHotel() {
        setNumHoteles(numHoteles + 1);
        setNumCasas(0);
        return titulo.getPrecioEdificar();
    }

    int hipotecar() {
        titulo.setHipotecada(Boolean.TRUE);
        return calcularValorHipoteca();
    }

    boolean propietarioEncarcelado() {
        return titulo.propietarioEncarcelado();
    }

    boolean sePuedeEdificarCasa(int factorEspeculador) {
        return numCasas < 4 * factorEspeculador;
    }

    boolean sePuedeEdificarHotel(int factorEspeculador) {
        return numHoteles < 4 * factorEspeculador && numCasas == 4 * factorEspeculador;
    }

    @Override
    boolean soyEdificable() {

        return true;
    }

    boolean estaHipotecada() {

        return titulo.getHipotecada();
    }

    boolean tengoPropietario() {
        return titulo.tengoPropietario();
    }

    int venderTitulo() {
        int precioCompra = coste + (numCasas + numHoteles) * titulo.getPrecioEdificar();
        int precioVenta = (int) (titulo.getFactorRevalorizacion() * precioCompra);

        titulo.setPropietario(null);
        setNumCasas(0);
        setNumHoteles(0);

        return precioVenta;
    }

    int precioTotalComprar() {
        throw new UnsupportedOperationException("Sin implementar");
    }

    @Override
    public String toString() {
        String s = super.toString();
        s += "\nCoste=" + coste
                + "\n\t" + titulo;
        if (titulo.tengoPropietario()) {
            s += "\n\tPropietario: " + titulo.getPropietario().getNombre();
        } else {
            s += "\n\tNo tiene propietario";
        }
        return s;
    }
}

File: QytetetRuby/lib/calle.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module ModeloQytetet
  class Calle < Casilla
    
    attr_accessor :num_hoteles, :num_casas, :titulo_propiedad
    
    def initialize(numero, coste, titulo)
      super(numero, coste, TipoCasilla::CALLE)
      @titulo_propiedad = titulo
      @num_hoteles = 0
      @num_casas = 0
      @titulo_propiedad.casilla = self
    end
    
    def to_s
      super + "Coste: #{@coste}\nNumero de Hoteles: #{@num_hoteles}" \
        "\nNumero de Casas: #{@num_casas} \n#{@titulo_propiedad}\n"
    end
    
    def asignar_propietario(jugador)
      @titulo_propiedad.propietario = jugador
      @titulo_propiedad
    end
    
    def calcular_valor_hipoteca
      hb = @titulo_propiedad.hipoteca_base
      (hb * (1 + num_casas * 0.5 + num_hoteles)).to_i
    end
    
    def cancelar_hipoteca
      @titulo_propiedad.hipotecada = false
      (calcular_valor_hipoteca * 1.1).to_i
    end
    
    def cobrar_alquiler
      coste_alquiler_base  = @titulo_propiedad.alquiler_base
      coste_alquiler_total = coste_alquiler_base +
        (@num_casas * 0.5 + @num_hoteles * 2).to_i

      @titulo_propiedad.cobrar_alquiler(coste_alquiler_total)
      coste_alquiler_total
    end
    
    def edificar_casa
      @num_casas += 1 
      get_precio_edificar
    end
    
    def edificar_hotel
      @num_hoteles += 1
      @num_casas = 0  
      get_precio_edificar
    end
    
    def esta_hipotecada
      @titulo_propiedad.hipotecada
    end
    
    def get_coste_hipoteca
      @titulo_propiedad.hipoteca_base
    end
    
    def get_precio_edificar
      @titulo_propiedad.precio_edificar
    end
    
    def hipotecar
      @titulo_propiedad.hipotecada = true
      calcular_valor_hipoteca
    end
    
    def precio_total_comprar
      precio_compra = @coste +
        (num_casas + num_hoteles) * @titulo_propiedad.precio_edificar
      (precio_compra + @titulo_propiedad.factor_rev * precio_compra).to_i
    end
    
    def propietario_encarcelado
      @titulo_propiedad.propietario_encarcelado
    end
    
    def se_puede_edificar_casa(factor_especulador)
      @num_casas < 4*factor_especulador
    end
    
    def se_puede_edificar_hotel(factor_especulador)
      @num_hoteles < 4*factor_especulador && @num_casas == 4*factor_especulador
    end
    
    
    def tengo_propietario
      @titulo_propiedad.tengo_propietario
    end
    
    def vender_titulo
      @titulo_propiedad.propietario = nil
      @num_hoteles = 0
      @num_casas = 0
      precio_total_comprar
    end
    
    def soy_edificable
      true
    end
  end
end
